framework/src/materials/Material.C

```diff
@@ -100,15 +102,25 @@ Material::Material(const InputParameters & parameters) :
 void
 Material::initStatefulProperties(unsigned int n_points)
 {
-  if (_has_stateful_property)
-    for (_qp = 0; _qp < n_points; ++_qp)
-      initQpStatefulProperties();
+  // checking for statefulness of properties via this loop is necessary
+  // because owned props might have been promoted to stateful by calls to
+  // getMaterialProperty[Old/Older] from other objects.  In these cases, this
+  // object won't otherwise know that it owns stateful properties.
+  for (auto& prop : _supplied_props)
+  {
+    if (_material_data->getMaterialPropertyStorage().isStatefulProp(prop))
+    {
+      for (_qp = 0; _qp < n_points; ++_qp)
+        initQpStatefulProperties();
+      return;
+    }
+  }
 }
 
 void
 Material::initQpStatefulProperties()
 {
-  mooseDoOnce(mooseWarning(std::string("Material \"") + name() + "\" declares one or more stateful properties but initQpStatefulProperties() was not overridden in the derived class."));
+  mooseDoOnce(mooseError(std::string("Material \"") + name() + "\" provides one or more stateful properties but initQpStatefulProperties() was not overridden in the derived class."));
```

Comment:
This function should not be called after properties before all non-stateful auto promoted of a material.